NullPointerException when building with newly-initialized, uncommitted git repo

Having just initialized my sandbox with a git repository I wanted to make sure that my project build before committing. This failed until I completed the repo initialization with a commit.

Relevant stack trace:
```
[ERROR] Failed to execute goal com.spotify:docker-maven-plugin:0.2.5:build (docker-build) on project docker-tomcat: Exception caught: NullPointerException -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.spotify:docker-maven-plugin:0.2.5:build (docker-build) on project docker-tomcat: Exception caught
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:216)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:120)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:355)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:155)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:584)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:216)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:160)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException: Exception caught
	at com.spotify.docker.AbstractDockerMojo.execute(AbstractDockerMojo.java:65)
	at com.spotify.docker.BuildMojo.execute(BuildMojo.java:79)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:132)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
	... 19 more
Caused by: java.lang.NullPointerException
	at com.spotify.docker.Git.getCommitId(Git.java:71)
	at com.spotify.docker.BuildMojo.execute(BuildMojo.java:209)
	at com.spotify.docker.AbstractDockerMojo.execute(AbstractDockerMojo.java:63)
	... 22 more
```

--- a/src/main/java/com/spotify/docker/DockerBuildInformation.java
+++ b/src/main/java/com/spotify/docker/DockerBuildInformation.java
@@ -28,6 +28,7 @@ import com.fasterxml.jackson.core.JsonProcessingException;
 import com.fasterxml.jackson.databind.ObjectMapper;
 
 import org.apache.maven.plugin.logging.Log;
+import org.eclipse.jgit.lib.ObjectId;
 import org.eclipse.jgit.lib.Repository;
 
 import java.io.IOException;
@@ -35,6 +36,7 @@ import java.io.IOException;
 import static com.fasterxml.jackson.annotation.JsonInclude.Include.NON_NULL;
 import static com.fasterxml.jackson.databind.MapperFeature.SORT_PROPERTIES_ALPHABETICALLY;
 import static com.fasterxml.jackson.databind.SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS;
+import static com.google.common.base.Strings.isNullOrEmpty;
 
 //* Might be overkill to do it this way, but should simplify things when if/when we add more to this
 public class DockerBuildInformation {
@@ -64,7 +66,10 @@ public class DockerBuildInformation {
       Repository repo = new Git().getRepo();
       if (repo != null) {
         this.repo   = repo.getConfig().getString("remote", "origin", "url");
-        this.commit = repo.resolve("HEAD").getName();
+        final ObjectId head = repo.resolve("HEAD");
+        if (head != null && !isNullOrEmpty(head.getName())) {
+          this.commit = head.getName();
+        }
       }
     } catch (IOException e) {
       log.error("Failed to read Git information", e);

--- a/src/main/java/com/spotify/docker/Git.java
+++ b/src/main/java/com/spotify/docker/Git.java
@@ -33,9 +33,11 @@ import org.eclipse.jgit.storage.file.FileRepositoryBuilder;
 
 import java.io.IOException;
 
+import static com.google.common.base.Strings.isNullOrEmpty;
+
 public class Git {
 
-  private final Repository repo;
+  private Repository repo;
 
   public Git() throws IOException {
     final FileRepositoryBuilder builder = new FileRepositoryBuilder();
@@ -55,6 +57,10 @@ public class Git {
     return repo;
   }
 
+  void setRepo(final Repository repo) {
+    this.repo = repo;
+  }
+
   public String getCommitId()
       throws GitAPIException, DockerException, IOException, MojoExecutionException {
 
@@ -68,6 +74,10 @@ public class Git {
     try {
       // get the first 7 characters of the latest commit
       final ObjectId head = repo.resolve("HEAD");
+      if (head == null || isNullOrEmpty(head.getName())) {
+        return null;
+      }
+
       result.append(head.getName().substring(0, 7));
       final org.eclipse.jgit.api.Git git = new org.eclipse.jgit.api.Git(repo);
 

--- a/src/test/java/com/spotify/docker/GitTest.java
+++ b/src/test/java/com/spotify/docker/GitTest.java
@@ -1,0 +1,76 @@
+/*
+ * Copyright (c) 2014 Spotify AB.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.docker;
+
+import com.spotify.docker.client.DockerException;
+
+import org.apache.maven.plugin.MojoExecutionException;
+import org.eclipse.jgit.api.errors.GitAPIException;
+import org.eclipse.jgit.lib.RefDatabase;
+import org.eclipse.jgit.lib.Repository;
+import org.junit.Before;
+import org.junit.Ignore;
+import org.junit.Test;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+import java.io.IOException;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.IsEqual.equalTo;
+import static org.mockito.Mockito.when;
+
+public class GitTest {
+
+  @Mock private Repository repo;
+  @Mock private RefDatabase refDatabase;
+
+  @Before
+  public void setUp() {
+    MockitoAnnotations.initMocks(this);
+  }
+
+  @Test(expected=MojoExecutionException.class)
+  public void testGetCommitIdInNonGitDirThrows()
+      throws IOException, DockerException, GitAPIException, MojoExecutionException {
+    final Git git = new Git();
+    git.setRepo(null);
+    git.getCommitId();
+  }
+
+  @Test
+  public void testGetCommitIdInGitDirWithNoCommitsReturnsNull()
+      throws IOException, DockerException, GitAPIException, MojoExecutionException {
+    when(repo.resolve("HEAD")).thenReturn(null);
+    final Git git = new Git();
+    git.setRepo(repo);
+    final String commitId = git.getCommitId();
+    assertThat(commitId, is(equalTo(null)));
+  }
+
+  @Ignore
+  public void testGetCommitId() {
+    // TODO (dxia) Refactor Git.java so that you can test Git.getCommitId() easily without having to
+    // mock objects two or more levels down.
+  }
+}

